### PR TITLE
Stats update

### PR DESCRIPTION
**Pull request by Suhail6inkling**
**CHANGES**
- Added the ".aoc stats" command to allow users to view their own or other's stats - including if and at what time parts were completed 

image

- Changed the original ".aoc stats" command to ".aoc daily" to avoid confusion

- Added ".aoc register" to allow users to link their usernames to their Discord accounts (for the use of ".aoc stats"

- Thus added a JSON file containing the ids of users along with their names (that can be read or written to)

### CODE REVIEW

File: bot/seasons/christmas/adventofcode.py [hunk 1]
```diff
@@ -413,20 +461,68 @@ def _build_about_embed(self) -> discord.Embed:
...
+                name=name,
+                value=value,
+                inline=False
+            )
+        return embed
+
+
+            
+
+
```
Comment (MarkKoz) [inline bot/seasons/christmas/adventofcode.py]: Too many blank lines here.

File: bot/seasons/christmas/adventofcode.py [hunk 2]
```diff
@@ -507,6 +613,19 @@ def __init__(self, members: list, owner_id: int, event_year: int):
 
         self.daily_completion_summary = self.calculate_daily_completion()
 
+    def get_position(self, member):
+        sorted_list = sorted(self.members, key = lambda x: x.local_score)
+        return sorted_list.index(member) + 1
+
+    def check_name(self, name):
+        return name in [member.name for member in self.members]
```
Comment (MarkKoz) [inline bot/seasons/christmas/adventofcode.py]:
```suggestion
        return any(name == member.name for member in self.members)
```

File: bot/seasons/christmas/adventofcode.py [hunk 3]
```diff
@@ -507,6 +613,19 @@ def __init__(self, members: list, owner_id: int, event_year: int):
...
+
+    def get_member(self, name):
+        matches = [member for member in self.members if member.name == name][0]
+        if matches:
+            return matches[0]
+        return None
```
Comment (MarkKoz) [inline bot/seasons/christmas/adventofcode.py]: Can't this whole function just be the following?
```py
return next((member for member in self.members if member.name == name), None)
```

File: bot/seasons/christmas/adventofcode.py [hunk 4]
```diff
@@ -507,6 +613,19 @@ def __init__(self, members: list, owner_id: int, event_year: int):
...
+        sorted_list = sorted(self.members, key = lambda x: x.local_score)
+        return sorted_list.index(member) + 1
+
+    def check_name(self, name):
+        return name in [member.name for member in self.members]
+
+    def get_member(self, name):
+        matches = [member for member in self.members if member.name == name][0]
+        if matches:
+            return matches[0]
```
Comment (MarkKoz) [inline bot/seasons/christmas/adventofcode.py]: It seems that getting index 0 of matches twice is a mistake.

File: bot/seasons/christmas/adventofcode.py [hunk 5]
```diff
@@ -446,8 +542,18 @@ def member_from_json(cls, injson: dict) -> "AocMember":
...
         )
 
+    @staticmethod
+    def _ts_from_int(unix: int) -> str:
+        """
+        Generates a String of the timestamp from a given Unix time
+        (Subtracts 18000 due to EST being UTC-5)
+        """
+        date = datetime.fromtimestamp(unix-18000)
+        return date.strftime("%d/%m/%Y %H:%M:%S")
```
Comment (MarkKoz) [inline bot/seasons/christmas/adventofcode.py]: I don't think everyone will agree on the use of this date format.

File: bot/seasons/christmas/adventofcode.py [hunk 6]
```diff
@@ -446,8 +542,18 @@ def member_from_json(cls, injson: dict) -> "AocMember":
...
+            full_dict = injson
         )
 
+    @staticmethod
+    def _ts_from_int(unix: int) -> str:
+        """
+        Generates a String of the timestamp from a given Unix time
+        (Subtracts 18000 due to EST being UTC-5)
+        """
+        date = datetime.fromtimestamp(unix-18000)
```
Comment (MarkKoz) [inline bot/seasons/christmas/adventofcode.py]: Why would you make times EST rather than UTC? Not everyone participating lives in eastern United States.

**Comment by sco1** [comment]
Other than @MarkKoz's comments, there are quite a few `flake8` issues that need resolving.

If this is going to return all 25 days worth of stats it might be better off if this was DMed to the invoking user otherwise the response is going to dominate the chat every time it's called (also an easy spam target without a cooldown).

Another casual musing would be to allow for an optional day specifier to limit the results.

**Comment by sco1** [comment]
Closing for inactivity. Feel free to ping us if this gets picked up again!

**Pull request closed without merging**